README.md
Arista Website
==============
This project tracks development on the Arista Transcoder website, located here:

 * http://www.transcoder.org/

Getting Started
---------------
Making changes to the site is easy once you get it set up:

    git clone git://github.com/danielgtaylor/arista-website.git
    cd arista-website
    ./init.sh
    ./manage.py runserver

Now go to http://localhost:8000/ and start playing around :-)



media/ChangeLog.txt
2011-06-20  Daniel G. Taylor <dan@programmer-art.org>

  * setup.py: Fail gracefully if byte-compiling doesn't work on install

  * arista-gtk: Remove debug print statement

  * locale/ar/LC_MESSAGES/arista.mo, locale/ast/LC_MESSAGES/arista.mo,
  locale/bg/LC_MESSAGES/arista.mo, locale/ca/LC_MESSAGES/arista.mo,
  locale/cs/LC_MESSAGES/arista.mo, locale/da/LC_MESSAGES/arista.mo,
  locale/de/LC_MESSAGES/arista.mo, locale/el/LC_MESSAGES/arista.mo,
  locale/en_GB/LC_MESSAGES/arista.mo, locale/es/LC_MESSAGES/arista.mo,
  locale/et/LC_MESSAGES/arista.mo, locale/eu/LC_MESSAGES/arista.mo,
  locale/fi/LC_MESSAGES/arista.mo, locale/fr/LC_MESSAGES/arista.mo,
  locale/gl/LC_MESSAGES/arista.mo, locale/hu/LC_MESSAGES/arista.mo,
  locale/ia/LC_MESSAGES/arista.mo, locale/id/LC_MESSAGES/arista.mo,
  locale/it/LC_MESSAGES/arista.mo, locale/ja/LC_MESSAGES/arista.mo,
  locale/jv/LC_MESSAGES/arista.mo, locale/kn/LC_MESSAGES/arista.mo,
  locale/lt/LC_MESSAGES/arista.mo, locale/nl/LC_MESSAGES/arista.mo,
  locale/pl/LC_MESSAGES/arista.mo, locale/pt/LC_MESSAGES/arista.mo,
  locale/pt_BR/LC_MESSAGES/arista.mo, locale/ro/LC_MESSAGES/arista.mo,
  locale/ru/LC_MESSAGES/arista.mo, locale/sk/LC_MESSAGES/arista.mo,
  locale/sl/LC_MESSAGES/arista.mo, locale/sr/LC_MESSAGES/arista.mo,
  locale/sv/LC_MESSAGES/arista.mo, locale/templates/arista.pot,
  locale/th/LC_MESSAGES/arista.mo, locale/tr/LC_MESSAGES/arista.mo,
  locale/uk/LC_MESSAGES/arista.mo, locale/zh_CN/LC_MESSAGES/arista.mo,
  locale/zh_TW/LC_MESSAGES/arista.mo: Update translations from Launchpad

2011-06-19  Daniel G. Taylor <dan@programmer-art.org>

  * arista-gtk: Fix output file naming when input path isn't the same as the
  output directory, also fixes naming and overwriting of inputs from dvd and
  v4l sources.

  * arista-gtk: Reorganize preset selection code based on mouse coordinates on
  the logo widget, change drag and drop to only work on the logo widget and
  preview widget and make it sensitive to the location that was dropped on,
  defaulting to WebM if no preset icon was dropped on.

2011-06-18  Daniel G. Taylor <dan@programmer-art.org>

  * README.md: Add link to translations pages on Launchpad so they are easier
  to find

  * arista-gtk: Modify logo widget to listen to button click events, have the
  main window listen for clicks and when one of the icons in the logo is
  clicked open the create conversion dialog with the corresponding preset
  selected. This should serve as a nice shortcut and help remove new user
  confusion as they may expect something to happen when clicking the icons in
  logo. Thanks to Primoz Verdnik for the user experience help.

2011-06-17  Daniel G. Taylor <dan@programmer-art.org>

  * arista/queue.py: Make sure to remove failed items from the queue after they
  error out, and set the pipe to not running so that the next item will start
  processing. Thanks to Bram Avontuur for pointing out this issue to me.

  * ui/main.ui: Change cancel button on main dialog to be the media stop button
  instead, as really we are stopping a job. The data transcoded so far will
  still work, and nothing is undone as cancel may imply. This also makes more
  sense when talking about live input sources that have no end. You aren't
  canceling them so much as stopping the conversion.

  * arista-gtk: Small cleanup - emove some old code that is no longer used

  * arista-gtk: Disable create button when no DVD is actually inserted,
  prevents accidentally starting DVD jobs and having them error out right away

  * arista/transcoder.py: Add tolerance value to audiorate element by default.
  This fixes some audio glitch issues, in particular from WMV input files.
  Fixes issue #12

  * presets/apple.json: Update Apple iOS presets to ease up on reference frames
  and set the x264 profile, which should limit the use of bframes and fix #79
  for GStreamer 0.10.32

2011-06-16  Daniel G. Taylor <dan@programmer-art.org>

  * arista-gtk: Implement basic drag and drop of single files and directories
  onto the main window. Multiple files currently does not work and requires
  larger changes.

  * arista-gtk: Fix icon size being used to lookup icons - makes file type
  icons and some other icons like the dvd disc show up properly in the source
  selector

  * arista-gtk: GTK GUI now runs reniced to +5 by default, which should allow
  other programs to take priority while transcodes are going on, making the
  system feel overall snappier on many different schedulers. Fixes issue #23.

  * arista-gtk: Handle bad values for icon in case of e.g. messed up tar file
  or user-modified json with errors

  * arista-gtk: Fix issue #7 - problem importing preset with apparently invalid
  value for icon

  * presets/playstation.json: Update PSP preset to force 30 frames per second,
  which is apparently required for some PSP versions according to issue #49

  * arista-gtk, help/ac-3.html, help/mp2.html, presets/dvd.json: Add MP2 and
  AC-3 audio formats, update DVD preset to use MP2 instead of MP3 as the audio
  format as this is what it should have used all along, whoops. Thanks to
  guiodic for the bug report.

2011-06-15  Daniel G. Taylor <dan@programmer-art.org>

  * help/style.css: Update help file style by adding slight rounded corners and
  text shadows

  * MANIFEST.in: Make sure to include CSS for help files when creating
  distributable archives for installation

  * locale/templates/arista.pot: Update translation template to include new
  preset reset functionality

  * utils/update_translation.sh: Fix small bug in translation update script

  * arista-gtk, arista-transcode, arista/presets.py, ui/prefs.ui: Add a method
  to completely reset default presets, add button to do so from GTK prefs
  dialog and an argument to do so via arista-transcode

  * arista/presets.py: Use set to find images when exporting a device preset so
  that a file doesn't get added to the archive multiple times, which causes
  issues when importing


urls.py
import settings

from django.conf.urls.defaults import *

# Uncomment the next two lines to enable the admin:
# from django.contrib import admin
# admin.autodiscover()

urlpatterns = patterns('',
    # Example:
    # (r'^website/', include('website.foo.urls')),

    # Uncomment the admin/doc line below and add 'django.contrib.admindocs' 
    # to INSTALLED_APPS to enable admin documentation:
    # (r'^admin/doc/', include('django.contrib.admindocs.urls')),

    # Uncomment the next line to enable the admin:
    # (r'^admin/', include(admin.site.urls)),
)

urlpatterns += patterns('main.views',
    (r'^$', 'about'),
    (r'^presets/$', 'presets'),
    (r'^presets/submit/$', 'preset_submit'),
    (r'^presets/create/$', 'preset_create'),
    (r'^screenshots/$', 'screenshots'),
    (r'^downloads/$', 'downloads'),
    (r'^contact/$', 'contact'),
    (r'^contact/thanks/$', 'thanks'),
)

if settings.DEBUG:
    urlpatterns += patterns('',
        # serving the admin media
        (r'^media/(?P<path>.*)$', 
            'django.views.static.serve', 
            {'document_root': settings.PROJECT_ROOT + '/media'}),
        (r'^media/admin/(?P<path>.*)$', 
            'django.views.static.serve', 
            {'document_root': settings.PROJECT_ROOT + '/media/admin'}),
    )



manage.py
#!/usr/bin/env python

import os
import sys

BASE = os.path.dirname(os.path.abspath(__file__))

# Path to where Django is
sys.path.insert(0, os.path.join(BASE, "contrib"))
for path in os.listdir(os.path.join(BASE, "contrib")):
    sys.path.insert(1, os.path.join(BASE, "contrib", path))

from django.core.management import execute_manager
try:
    import settings # Assumed to be in the same directory.
except ImportError:
    import sys
    sys.stderr.write("Error: Can't find the file 'settings.py' in the directory containing %r. It appears you've customized things.\nYou'll have to run django-admin.py, passing it your settings module.\n(If the file settings.py does indeed exist, it's causing an ImportError somehow.)\n" % __file__)
    sys.exit(1)

if __name__ == "__main__":
    execute_manager(settings)


main/views.py
# Create your views here.

try:
    import json
except ImportError:
    import simplejson as json

import datetime
import glob
import os
import StringIO
import tarfile

from django.shortcuts import render_to_response
from django.template import RequestContext
from django.core.mail import send_mail, EmailMessage
from django.http import HttpResponse, HttpResponseRedirect

import settings

from forms import ContactForm, SubmitForm, PresetForm

def render(request, template, params = {}, mimetype = 'text/html'):
    """
        Render a response with the proper request context so that all
        variables expected in templates are indeed there from installed
        apps and middleware.
    """
    return render_to_response(template, params, mimetype = mimetype,
                               context_instance = RequestContext(request))

def about(request):
    return render(request, "about.html")

def presets(request):
    presets = []
    
    search_path = os.path.join(settings.MEDIA_ROOT, "presets", "*.json")
    
    for filename in glob.glob(search_path):
        presets.append(json.load(open(filename)))
        presets[-1]["name"] = os.path.basename(filename)[:-5]
        if presets[-1]["make"] != "Generic":
            presets[-1]["display_name"] = presets[-1]["make"] + " " + presets[-1]["model"]
        else:
            presets[-1]["display_name"] = presets[-1]["model"]

    return render(request, "presets/list.html", {
        "presets": sorted(presets, lambda x,y: cmp(x["display_name"], y["display_name"])),
    })

def preset_submit(request):
    if request.method == "POST":
        form = SubmitForm(request.POST, request.FILES)
        
        if form.is_valid():
            preset = request.FILES["preset_file"]
            message = EmailMessage(**{
                "subject": "[Arista] Preset for %s" % preset.name[:-8],
                "body": "A user has uploaded a new preset!",
                "from_email": form.cleaned_data["email"],
                "to": [x[1] for x in settings.ADMINS],
                "attachments": [
                    (preset.name, preset.read(), "application/x-bzip-compressed-tar"),
                ],
            })
            
            message.send()
        
            return HttpResponseRedirect("/presets/?thanks=1")
    else:
        form = SubmitForm()
    
    return render(request, "presets/submit.html", {
        "form": form,
    })

def preset_create(request):
    if request.method == "POST":
        form = PresetForm(request.POST, request.FILES)
        
        if form.is_valid():
            response = HttpResponse(mimetype="application/x-bzip-compressed-tar")
            response["Content-Disposition"] = "attachment; filename=%s.tar.bz2" % form.cleaned_data["short_name"]
            
            tar = tarfile.open(fileobj=response, mode="w:bz2")
            
            json = StringIO.StringIO()
            json.write(form.preset_json.encode("utf-8"))
            json.seek(0)
            info = tarfile.TarInfo(name=form.cleaned_data["short_name"] + ".json")
            info.size = len(json.getvalue())
            tar.addfile(info, fileobj=json)
            
            icon = StringIO.StringIO()
            icon.write(request.FILES["icon"].read())
            icon.seek(0)
            info = tarfile.TarInfo(name=request.FILES["icon"].name)
            info.size = request.FILES["icon"].size
            tar.addfile(info, fileobj=icon)
            
            tar.close()
            
            return response
    else:
        form = PresetForm()
    
    return render(request, "presets/create.html", {
        "form": form,
    })

def screenshots(request):
    return render(request, "screenshots.html")

def downloads(request):
    data = json.loads(open(os.path.join(settings.MEDIA_ROOT, "downloads.json")).read())
    
    for release in data:
        release[1] = datetime.datetime.fromtimestamp(release[1])
    
    return render(request, "downloads.html", {
        "data": data,
    })

def contact(request):
    if request.method == "POST":
        form = ContactForm(request.POST)
        
        if form.is_valid():
            send_mail("[Arista] " + form.cleaned_data["title"], form.cleaned_data["message"], form.cleaned_data["email"], [x[1] for x in settings.ADMINS])
            
            return HttpResponseRedirect("/contact/thanks")
    else:
        form = ContactForm()

    return render(request, "contact.html", {
        "form": form,
    })

def thanks(request):
    return render(request, "thanks.html")



main/management/commands/presetcache.py
try:
    import json
except ImportError:
    import simplejson as json

import glob
import os
import subprocess

from django.core.management.base import BaseCommand, CommandError

import settings

class Command(BaseCommand):
    args = '[device_short_name]'
    help = 'Update the presets cache (tar archives and latest.txt)'
    search_path = os.path.join(settings.MEDIA_ROOT, "presets")

    def handle(self, *args, **options):
        if len(args):
            device = json.load(open(os.path.join(self.search_path, args[0] + ".json")))
            self.gentar(args[0], device)
        else:
            # All files
            latest = []
            
            for filename in glob.glob(os.path.join(self.search_path, "*.json")):
                device = json.load(open(filename))
                short_name = os.path.basename(filename)[:-5]
                latest.append("%s, %s" % (short_name, device["version"]))
                self.gentar(short_name, device)
            
            open(os.path.join(self.search_path, "latest.txt"), "w").write("\n".join(sorted(latest)))
    
    def gentar(self, short_name, device):
        images = []
        
        if "icon" in device:
            images.append(device["icon"][7:])
        
        for preset in device["presets"]:
            if "icon" in preset:
                images.append(preset["icon"][7:])
        
        os.chdir(self.search_path)
        archive = short_name + ".tar.bz2"
        files = " ".join([short_name + ".json"] + images)
        cmd = "tar -cjf %s %s" % (archive, files)
        print "Creating %s (version %s)" % (archive, device["version"])
        subprocess.call(cmd, shell=True)   



main/management/commands/downloadscache.py
try:
    import json
except ImportError:
    import simplejson as json

import os
import time

from django.core.management.base import BaseCommand, CommandError

from github2.client import Github

import settings

class Command(BaseCommand):
    args = ''
    help = 'Update the downloads cache (used to render the downloads page)'

    def handle(self, *args, **options):
        github = Github()
        
        data = []
        
        tags = github.repos.tags("danielgtaylor/arista")
        
        for tag_name in sorted(tags.keys(), reverse=True):
            commit = github.commits.show("danielgtaylor/arista", tags[tag_name])
            data.append([tag_name, time.mktime(commit.authored_date.timetuple()), commit.id])
        
        open(os.path.join(settings.MEDIA_ROOT, "downloads.json"), "w").write(json.dumps(data, indent=4))
        
        print "Downloads cache updated."



main/forms.py
import os

from django import forms
from django.forms import widgets

import settings

REGEX_RANGE = "[0-9]+,\s?[0-9]+"
REGEX_RANGE_FLOAT = "[0-9.]+,\s?[0-9.]+"

class ContactForm(forms.Form):
    email = forms.EmailField()
    title = forms.CharField()
    message = forms.CharField(widget=widgets.Textarea())

class SubmitForm(forms.Form):
    email = forms.EmailField()
    preset_file = forms.FileField()

class PresetForm(forms.Form):
    template = unicode(open(os.path.join(settings.PROJECT_ROOT, "media", "template.json")).read())

    author_name = forms.CharField()
    author_email = forms.EmailField()
    short_name = forms.CharField(help_text="A short, one-word, unique identifier")
    make = forms.CharField(help_text="The manufacturer of this device, e.g. 'Apple'", initial="Generic")
    model = forms.CharField(help_text="The model of this device, e.g. 'iPhone'")
    description = forms.CharField(help_text="A short description, e.g. 'H.264/AAC in MP4 for iPhones")
    icon = forms.FileField(help_text="An SVG or PNG file")
    container = forms.ChoiceField(choices = [
        ("qtmux", "MP4"),
        ("webmmux", "WebM"),
        ("matroskamux", "Matroska"),
        ("avimux", "AVI"),
        ("mpegtsmux", "MPEG-TS"),
        ("ffmux_dvd", "DVD"),
        ("oggmux", "Ogg"),
    ])
    video_codec = forms.ChoiceField(choices = [
        ("x264enc", "H.264"),
        ("vp8enc", "VP8"),
        ("xvidenc", "MPEG4"),
        ("mpeg2enc", "MPEG2 DVD"),
        ("theoraenc", "Theora"),
    ])
    video_quality = forms.ChoiceField(choices = [
        ("worst", "Worst (small files)"),
        ("worse", "Worse"),
        ("good", "Good"),
        ("better", "Better"),
        ("best", "Best (large files)"),
    ], help_text="Visual quality where bitrate is automatically determined", initial="good")
    video_bitrate = forms.IntegerField(help_text="The video bitrate in kbps", initial=4000)
    width = forms.RegexField(regex=REGEX_RANGE, help_text="Range of allowed widths, like 'min, max'", initial="120, 1280")
    height = forms.RegexField(regex=REGEX_RANGE, help_text="Range of allowed heights, like 'min, max'", initial="120, 720")
    framerate = forms.RegexField(regex=REGEX_RANGE_FLOAT, help_text="Range of allowed framerates, like 'min, max'", initial="1, 30")
    audio_codec = forms.ChoiceField([
        ("faac", "AAC"),
        ("lame", "MP3"),
        ("vorbisenc", "Vorbis"),
        ("ffenc_ac3", "AC3"),
    ])
    audio_bitrate = forms.IntegerField(help_text="The audio bitrate in kbps", initial=128)
    audio_channels = forms.RegexField(regex=REGEX_RANGE, help_text="Range of allowed channels, like 'min, max'", initial="1, 2")
    
    def clean_icon(self):
        data = self.cleaned_data["icon"]
        
        if data.size > 1024 ** 2:
            raise forms.ValidationError("File size must be under 1 MiB")
        
        return data
    
    @property
    def extension(self):
        return {
            "qtmux": "mp4",
            "matroskamux": "mkv",
            "avimux": "avi",
            "ffmux_dvd": "mpg",
            "oggmux": "ogg",
            "webmmux": "webm",
        }.get(self.cleaned_data["container"], "mp4")
    
    @property
    def x264_video_quality(self):
        return {
            "worst": 27.0,
            "worse": 24.0,
            "good": 21.0,
            "better": 19.0,
            "best": 18.0,
        }.get(self.cleaned_data["video_quality"], 21.0)
    
    @property
    def vp8_video_quality(self):
        return {
            "worst": 2,
            "worse": 4,
            "good": 5,
            "better": 6,
            "best": 8,
        }.get(self.cleaned_data["video_quality"], 5)
    
    @property
    def xvid_video_quality(self):
        return {
            "worst": 10.0,
            "worse": 8.0,
            "good": 5.0,
            "better": 3.0,
            "best": 2.0,
        }.get(self.cleaned_data["video_quality"], 5)
    
    @property
    def theora_video_quality(self):
        return {
            "worst": 15,
            "worse": 30,
            "good": 40,
            "better": 60,
            "best": 80,
        }.get(self.cleaned_data["video_quality"], 40)
    
    @property
    def video_pass(self):
        cleaned = self.cleaned_data
        
        return {
            "x264enc": "pass=qual quantizer=%d subme=6 cabac=0 threads=0" % self.x264_video_quality,
            "vp8enc": "quality=%d threads=%%(threads)s speed=2" % self.vp8_video_quality,
            "xvidenc": "pass=quant quantizer=%d max-bframes=0 trellis=true" % self.xvid_video_quality,
            "mpeg2enc": "format=9 bitrate=%s" % self.cleaned_data["video_bitrate"],
            "theoraenc": "border=0 quality=%d keyframe-freq=30" % self.theora_video_quality,
        }.get(cleaned["video_codec"], "")
    
    @property
    def audio_pass(self):
        cleaned = self.cleaned_data
        
        return {
            "faac": "bitrate=%d profile=LC" % (int(cleaned["audio_bitrate"]) * 1024),
            "lame": "bitrate=%s" % cleaned["audio_bitrate"],
            "vorbisenc": "bitrate=%d" % (int(cleaned["audio_bitrate"]) * 1024),
            "ffenc_ac3": "bitrate=%d" % (int(cleaned["audio_bitrate"]) * 1024),
        }.get(cleaned["audio_codec"], "")
    
    @property
    def audio_container(self):
        cleaned = self.cleaned_data
        return {
            "lame": "",
        }.get(cleaned["audio_codec"], cleaned["container"])
    
    @property
    def preset_json(self):
        cleaned = self.cleaned_data
        return self.template % {
            "make": cleaned["make"],
            "model": cleaned["model"],
            "description": cleaned["description"],
            "version": "1.0",
            "author_name": cleaned["author_name"],
            "author_email": cleaned["author_email"],
            "icon": cleaned["icon"].name,
            "extension": self.extension,
            "container": cleaned["container"],
            "video_name": cleaned["video_codec"],
            "video_container": cleaned["container"],
            "width": cleaned["width"],
            "height": cleaned["height"],
            "rate": cleaned["framerate"],
            "video_pass": self.video_pass,
            "audio_name": cleaned["audio_codec"],
            "audio_container": self.audio_container,
            "audio_pass": self.audio_pass,
            "audio_channels": cleaned["audio_channels"],
        }
    

